shuup/xtheme/plugins/products_async.py

```diff
@@ -0,0 +1,290 @@
+# -*- coding: utf-8 -*-
+# This file is part of Shuup.
+#
+# Copyright (c) 2012-2020, Shoop Commerce Ltd. All rights reserved.
+#
+# This source code is licensed under the OSL-3.0 license found in the
+# LICENSE file in the root directory of this source tree.
+from django import forms
+from django.urls import reverse
+from django.utils.translation import ugettext_lazy as _
+from enumfields import Enum
+
+from shuup.core.models import Product, ProductCrossSell, ProductCrossSellType
+from shuup.front.template_helpers.general import (
+    get_best_selling_products, get_newest_products,
+    get_products_for_categories, get_random_products
+)
+from shuup.front.template_helpers.product import (
+    get_product_cross_sells, map_relation_type
+)
+from shuup.xtheme import TemplatedPlugin
+from shuup.xtheme.plugins.forms import GenericPluginForm, TranslatableField
+from shuup.xtheme.plugins.widgets import (
+    XThemeSelect2ModelChoiceField, XThemeSelect2ModelMultipleChoiceField
+)
+
+
+class HighlightType(Enum):
+    NEWEST = "newest"
+    BEST_SELLING = "best_selling"
+    RANDOM = "random"
+
+    class Labels:
+        NEWEST = _("Newest")
+        BEST_SELLING = _("Best Selling")
+        RANDOM = _("Random")
+
+
+class ProductHighlightPlugin(TemplatedPlugin):
```

Comment:
unfortunately, we are repeating ourselves with all this duplicated code. What is we have some sort of async field that is disabled by default that would then change the way we render the template and the context is gethered?